File: apps/cli/lib/commands/list.ts
import chalk from "chalk";
import ora from "ora";
import { requireAuth } from "../auth.js";
import { getSites } from "../api-client.js";
import { displayError, formatDate, getSiteUrl } from "../utils.js";
import { API_URL } from "../const.js";
import { capture, flushTelemetry, CLI_VERSION } from "../telemetry.js";

/**
 * List command - show all sites for the authenticated user
 */
export async function listCommand(): Promise<void> {
  const startTime = Date.now();
  capture("command_started", { command: "list", cli_version: CLI_VERSION });
  try {
    const user = await requireAuth();

    const spinner = ora("Fetching sites...").start();

    // Get sites from API
    const sitesData = await getSites();
    const sites = sitesData.sites || [];

    spinner.stop();

    if (sites.length === 0) {
      console.log(chalk.gray("\nNo sites found.\n"));
      return;
    }

    console.log(chalk.bold(`\nFound ${sites.length} site(s):\n`));

    for (const site of sites) {
      const url = getSiteUrl(
        site.projectName,
        user.username || user.email || "user",
      );
      const dashboardUrl = `${API_URL}/site/${site.id}/settings`;
      console.log(chalk.cyan(`  ${site.projectName}`));
      console.log(chalk.gray(`    URL: ${url}`));
      console.log(chalk.gray(`    Dashboard URL: ${dashboardUrl}`));
      console.log(chalk.gray(`    Created: ${formatDate(site.createdAt)}`));
      console.log(
        chalk.gray(
          `    Updated: ${formatDate(site.updatedAt || site.createdAt)}`,
        ),
      );
      console.log();
    }
    capture("command_succeeded", {
      command: "list",
      cli_version: CLI_VERSION,
      duration_ms: Date.now() - startTime,
    });
  } catch (error) {
    capture("command_failed", {
      command: "list",
      cli_version: CLI_VERSION,
      duration_ms: Date.now() - startTime,
      error_type: error instanceof Error ? error.constructor.name : "Unknown",
      error_message: error instanceof Error ? error.message : String(error),
    });
    if (error instanceof Error) {
      displayError(error.message);
      console.error(chalk.gray(error.stack));
    } else {
      displayError("An unknown error occurred");
    }
    await flushTelemetry();
    process.exit(1);
  } finally {
    await flushTelemetry();
  }
}

File: apps/cli-go/cmd/list.go
package cmd

import (
	"fmt"
	"time"

	"github.com/flowershow/publish/internal/api"
	"github.com/flowershow/publish/internal/auth"
	"github.com/flowershow/publish/internal/config"
	"github.com/flowershow/publish/internal/telemetry"
	"github.com/flowershow/publish/internal/ui"
	"github.com/spf13/cobra"
)

var listCmd = &cobra.Command{
	Use:   "list",
	Short: "List all published sites",
	Args:  cobra.NoArgs,
	RunE: func(cmd *cobra.Command, args []string) error {
		ui.Header("List Sites")
		return runList()
	},
}

func init() {
	rootCmd.AddCommand(listCmd)
}

func runList() error {
	startTime := time.Now()
	telemetry.Capture("command_started", map[string]interface{}{
		"command":     "list",
		"cli_version": config.Version,
	})
	defer func() { telemetry.Flush() }()

	// Authenticate
	tokenData, err := auth.GetToken()
	if err != nil || tokenData == nil {
		ui.PrintError("You must be authenticated to use this command.\nRun `publish auth login` to authenticate.")
		return nil
	}
	if _, err := auth.GetUserInfo(config.APIURL(), tokenData.Token); err != nil {
		ui.PrintError("You must be authenticated to use this command.\nRun `publish auth login` to authenticate.")
		return nil
	}

	sp := ui.NewSpinner()
	sp.Start("Fetching sites...")

	sitesData, err := api.GetSites()
	if err != nil {
		sp.Fail("Failed to fetch sites")
		ui.PrintError(err.Error())
		telemetry.Capture("command_failed", map[string]interface{}{
			"command":       "list",
			"cli_version":   config.Version,
			"duration_ms":   time.Since(startTime).Milliseconds(),
			"error_message": err.Error(),
		})
		return nil
	}
	sp.Stop()

	sites := sitesData.Sites
	if len(sites) == 0 {
		fmt.Printf("\n%s\n\n", ui.Gray("No sites found."))
		return nil
	}

	fmt.Printf("\n%s\n\n", ui.Bold(fmt.Sprintf("Found %d site(s):", len(sites))))

	for _, site := range sites {
		dashboardURL := fmt.Sprintf("%s/site/%s/settings", config.APIURL(), site.ID)
		updatedAt := site.CreatedAt
		if site.UpdatedAt != nil {
			updatedAt = *site.UpdatedAt
		}
		fmt.Printf("  %s\n", ui.Cyan(site.ProjectName))
		fmt.Printf("    %s\n", ui.Gray("URL: "+site.URL))
		fmt.Printf("    %s\n", ui.Gray("Dashboard URL: "+dashboardURL))
		fmt.Printf("    %s\n", ui.Gray("Created: "+formatDisplayDate(site.CreatedAt)))
		fmt.Printf("    %s\n", ui.Gray("Updated: "+formatDisplayDate(updatedAt)))
		fmt.Println()
	}

	telemetry.Capture("command_succeeded", map[string]interface{}{
		"command":     "list",
		"cli_version": config.Version,
		"duration_ms": time.Since(startTime).Milliseconds(),
	})
	return nil
}

func formatDisplayDate(dateStr string) string {
	t, err := time.Parse(time.RFC3339, dateStr)
	if err != nil {
		return dateStr
	}
	return t.Local().Format("2006-01-02 15:04:05")
}
